### ISSUE TITLE
[clang-tidy] `bugprone-unused-return-value` false positive for `unique_ptr` assignment

### ISSUE THREAD

**Issue by zufuliu** [llvm/llvm-project#84314]
https://godbolt.org/z/a6f19rboW

```c++
// clang-tidy --checks=bugprone-unused-return-value test.cpp --
#include <cstdlib>
#include <memory>

extern std::unique_ptr<int> alloc();

struct Foo {
    std::unique_ptr<int> foo;
    std::unique_ptr<int> bar;
    void *baz;
    Foo() {
        foo = std::make_unique<int>();
        bar = alloc();
        baz = malloc(sizeof(int));
    }
};

void f1(std::unique_ptr<int> &foo) {
    foo = std::make_unique<int>();
}

void f2(std::unique_ptr<int> &bar) {
    bar = alloc();
}

void f3(void* &baz) {
    baz = malloc(sizeof(int));
}
```

```
--checks=bugprone-unused-return-value
[<source>:12:9: warning: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value]](javascript:;)
   12 |         foo = std::make_unique<int>();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[<source>:13:9: warning: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value]](javascript:;)
   13 |         bar = alloc();
      |         ^~~~~~~~~~~~~
[<source>:19:5: warning: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value]](javascript:;)
   19 |     foo = std::make_unique<int>();
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[<source>:23:5: warning: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value]](javascript:;)
   23 |     bar = alloc();
      |     ^~~~~~~~~~~~~
4 warnings gener
```

**Comment by chrchr-github** [comment]
This is a false **positive**, right?

**Comment by HerrCai0907** [comment]
I think it is false negative, This check should not detect `malloc` and `std::unique_ptr`

**Comment by EugeneZelenko** [comment]
@HerrCai0907: false negative means that something is not detected when it should be.

**Comment by firewave** [comment]
It seems like this is a bigger issue.

If you use `::.*` to match all functions it will report the warning on *every* assignment (`=` and `+=`) regardless of type.

**Comment by firewave** [comment]
Here's an example:

```cpp
#include <string>

class C {
public:
    void setS(const char* s) {
        mS = s;
    }
private:
    std::string mS;
};
```

```
test.cpp:6:9: error: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value,-warnings-as-errors]
    6 |         mSt = s;
      |         ^~~~~~~
```

(No godbolt link as I could not figure out how to make the passing of the option work)
